### PR TITLE
Router .parameter method implementation

### PR DESCRIPTION
**Pull request by naithar**
Another try to implement a `.parameter` method for `Router` to handle parsed parameters.
Previous implementation: IBM-Swift/Kitura#480

## Motivation and Context
IBM-Swift/Kitura#295

## How Has This Been Tested?
Added test to `TestRequests.swift` file

## Checklist:

- [x] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [ ] If applicable, I have updated the documentation accordingly.
- [x] If applicable, I have added tests to cover my changes.

### CODE REVIEW

File: Sources/Kitura/Router.swift [hunk 1]
```diff
@@ -157,6 +160,31 @@ public class Router {
         self.all(route, middleware: subrouter)
         return subrouter
     }
+
+    ///
```
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: Missing API documentation here

File: Sources/Kitura/Router.swift [hunk 2]
```diff
@@ -157,6 +160,31 @@ public class Router {
...
+    @discardableResult
+    public func parameter(_ name: String, handler: @escaping RouterParameterHandler...) -> Router {
+        return self.parameter([name], handlers: handler)
+    }
+
+    ///
```
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: Missing API documentation here

File: Sources/Kitura/Router.swift [hunk 3]
```diff
@@ -157,6 +160,31 @@ public class Router {
...
+    @discardableResult
+    public func parameter(_ names: [String], handler: @escaping RouterParameterHandler...) -> Router {
+        return self.parameter(names, handlers: handler)
+    }
+
+    ///
```
Comment (shmuelk) [inline Sources/Kitura/Router.swift]: Missing API documentation here

File: Sources/Kitura/RouterParameterWalker.swift [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    init(handlers: [String : [RouterParameterHandler]]) {
+        self.parameterHandlers = handlers
+    }
+
+    func handle(request: RouterRequest, response: RouterResponse, with callback: @escaping () -> Void) {
+        let filtered = request.parameters.filter { (key, _) in self.parameterHandlers.keys.contains(key) }
+        self.handle(filtered: filtered, request: request, response: response, with: callback)
+    }
+
```
Comment (shmuelk) [inline Sources/Kitura/RouterParameterWalker.swift]: It took a while to understand what was going on here.

If I understand correctly, you are modifying the copied arrays of parameters and parameterHandlers in order to do the recursion and/or looping without an index (i.e. other state in the RouterParameterWalkser class.

I have two concerns here:
    1. Performance. Which performs better the algorithm in the RouterElementWalker class, which uses an index, albeit captures a local variable or this one which is mdifying, thus copying the arrays.
    2. Consistency. I would have liked all of the "walker" classes to be similar in nature. It makes it easier to maintain them.

**Comment by shmuelk** [comment]
In addition to the changes requested, please update your branch from master.....

**Comment by naithar** [comment]
@shmuelk updated based on review

File: Sources/Kitura/RouterParameterWalker.swift [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    /// Collection of 'RouterParameterHandler' for specified parameter name
+    private var parameterHandlers: [String : [RouterParameterHandler]]
+
+    init(handlers: [String : [RouterParameterHandler]]) {
+        self.parameterHandlers = handlers
+    }
+
+    /// Invoke all possible parameter handlers for request
+    ///
+    /// - Parameter request: A current `RouterRequst` that is handled by a server
```
Comment (ianpartridge) [inline Sources/Kitura/RouterParameterWalker.swift]: Typo - `RouterRequest`

File: Sources/Kitura/RouterParameterWalker.swift [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    /// - Parameter response: A current `RouterResponse` that is handled by a server
+    /// - Parameter callback: A callback that will be invoked when all handlers are invoked
```
Comment (ianpartridge) [inline Sources/Kitura/RouterParameterWalker.swift]: Should this be "**after** all handlers are invoked"?

File: Sources/Kitura/Router.swift [hunk 7]
```diff
@@ -53,6 +53,9 @@ public class Router {
     /// Flag to enable/disable access to parent router's params
     private let mergeParameters: Bool
 
+    ///
```
Comment (ianpartridge) [inline Sources/Kitura/Router.swift]: Documentation needed.

File: Sources/Kitura/Router.swift [hunk 8]
```diff
@@ -157,6 +160,46 @@ public class Router {
         self.all(route, middleware: subrouter)
         return subrouter
     }
+
+    /// Handle an URL parameter
```
Comment (ianpartridge) [inline Sources/Kitura/Router.swift]: Can these comments please explain WHY someone would use this functionality? What is the purpose of the feature / what problem does it solve?

File: Sources/Kitura/RouterElementWalker.swift [hunk 9]
```diff
@@ -36,12 +36,18 @@ class RouterElementWalker {
     /// Copy of request parameters to enable mergeParams
     private let parameters: [String: String]
 
-    init(elements: [RouterElement], request: RouterRequest, response: RouterResponse, callback: @escaping () -> Void) {
+    ///
```
Comment (ianpartridge) [inline Sources/Kitura/RouterElementWalker.swift]: Comment needed.

File: Sources/Kitura/RouterParameterWalker.swift [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            callback()
+            return
+        }
+
+        let filtered = request.parameters.filter { (key, _) in self.parameterHandlers.keys.contains(key) }
+        self.handle(filtered: filtered, request: request, response: response, with: callback)
+    }
+
+    private func handle(filtered: [(String, String)], request: RouterRequest, response: RouterResponse, with callback: @escaping () -> Void) {
+        if filtered.count > 0 {
```
Comment (ianpartridge) [inline Sources/Kitura/RouterParameterWalker.swift]: I would use the same pattern as in the other `handle()` function:

```swift
guard filtered.count > 0 else {
    callback()
    return
}

/// rest of function
```

File: Tests/KituraTests/TestRequests.swift [hunk 11]
```diff
@@ -26,7 +26,8 @@ class TestRequests: XCTestCase {
         return [
                    ("testURLParameters", testURLParameters),
                    ("testCustomMiddlewareURLParameter", testCustomMiddlewareURLParameter),
-                   ("testCustomMiddlewareURLParameterWithQueryParam", testCustomMiddlewareURLParameterWithQueryParam)
+                   ("testCustomMiddlewareURLParameterWithQueryParam", testCustomMiddlewareURLParameterWithQueryParam),
+                   ("testParameters", testParameters)
```
Comment (ianpartridge) [inline Tests/KituraTests/TestRequests.swift]: You can leave a trailing comma here so future tests can be added without modifying this line.

File: Tests/KituraTests/TestRequests.swift [hunk 12]
```diff
@@ -123,4 +124,40 @@ class TestRequests: XCTestCase {
 
         return router
     }
+
+    func testParameters() {
```
Comment (ianpartridge) [inline Tests/KituraTests/TestRequests.swift]: Can we expand the tests a little more? I feel like this is just the basic "hello world" test and there must be more complicated examples we should cover as well.

**Comment by naithar** [comment]
@ianpartridge updated comments and added tests. Not sure if there is more tests to add.

**Comment by naithar** [comment]
There is also another thing i would like to make clear for myself. Currently `RouterParameterWalker` is performed after `RouterElementWalker` performs all it's handlers. [This commit](https://github.com/naithar/Kitura/commit/638811974f031db24bafdce49027d33c91dd54e0) fixes this, but is it actually an issue to be addressed?

**Comment by shmuelk** [comment]
@naithar Do you mean The RouterElementWalker invokes the RouterParameterWalker after all of the routes have been processed (checked for match.....) ?

If so, I think that would be a problem. Every time a RouterElement matches a path, the set of request parameters gets reset to just those in the request. The set may also include parameters from parent routers  if the mergeParameters option was set.

I guess the RouterParameterWalkser needs to be invoked every time a RouterElement has a match and invokes the RouterMiddlewareWalker.

**Comment by naithar** [comment]
@shmuelk currently `RouterParameterWalker` called before parameters are reset, but after element and it's middlewares is processed. In branch that i linked earlier (issue_295-2), i moved `RouterParameterWalker` call right after parameters have been set but before middlewares start handling.

**Comment by naithar** [comment]
I can merge it to this PR for review.

**Comment by shmuelk** [comment]
@naithar That's what I thought I saw in the code. But I wasn't sure after your comment.

The real issue between calling the RouterParameterWalkser before the handlers get called, is what happens when people don't cal the various callbacks. In theory in a RouteHandler, one doesn't have to call net(), as long as one calls end() on the response. The same is true if one doesn't call the callback in a RouterParameterHandler. Maybe that's just life....

**Comment by naithar** [comment]
[This test](https://github.com/naithar/Kitura/blob/638811974f031db24bafdce49027d33c91dd54e0/Tests/KituraTests/TestRequests.swift#L208-L220) is made to check the correct response handling when user doesn't call `next` in `RouterParameterHandler`, but calls `.end()` on response. If i'm not mistaken the result should be the same as replacing it with just:

```swift
        router.get("users/:user/:id") { request, response, next in
            XCTAssertNotNil(request.parameters["id"])
            guard Int(request.parameters["id"]!) != nil else {
                try response.status(.notAcceptable).end()
                return
            }

            response.headers["User-Id"] = request.parameters["id"]
            response.status(.OK).send(data: "\(request.parameters["id"]!)".data(using: .utf8)!)
            next()
        }
```

Which is checked and working.

**Comment by shmuelk** [comment]
@naithar Excellent.

**Comment by naithar** [comment]
@shmuelk merged. Travis Linux build is currently failing, but it's not related to PR as far as i understand

**Comment by shmuelk** [comment]
@naithar We know about the Travis-CI problem. I reported it to them.

**Pull request merged**